### PR TITLE
mesh: Wire ComputedRoutes into the ProxyStateTemplate via the sidecar controller

### PR DESCRIPTION
**Pull request by rboyer**
### Description

Reworks the sidecar controller to accept `ComputedRoutes` as an input and use it to generate appropriate ProxyStateTemplate resources containing L4/L7 mesh configuration.

TODO

- [x] finish dangling mapper work (see comment)
~- [ ] wire up `FailoverPolicy` as aggregate clusters into IR~ (spun out into a new ticket)
- [ ] more tests
- [x] remove some now-deprecated code

NET-5348, NET-5210, NET-5064

### CODE REVIEW

File: proto-public/pbmesh/v1alpha1/http_route_retries.proto
Comment (rboyer) [inline proto-public/pbmesh/v1alpha1/http_route_retries.proto:12]: This was on my list to do and just did it in passing.

File: proto-public/pbmesh/v1alpha1/pbproxystate/route.proto
Comment (rboyer) [inline proto-public/pbmesh/v1alpha1/pbproxystate/route.proto:133]: Moved to `cluster.proto`

File: proto-public/pbmesh/v1alpha1/pbproxystate/cluster.proto
Comment (rboyer) [inline proto-public/pbmesh/v1alpha1/pbproxystate/cluster.proto:80]: Moved from `route.proto`

File: proto-public/pbmesh/v1alpha1/pbproxystate/listener.proto
Comment (rboyer) [inline proto-public/pbmesh/v1alpha1/pbproxystate/listener.proto:93]: This allows for `TCPRoute` to be direct or to do splits.

File: proto-public/pbmesh/v1alpha1/computed_routes.proto
Comment (rboyer) [inline proto-public/pbmesh/v1alpha1/computed_routes.proto:20]: The changes here were ergonomic while I tried to actually use the new resource.

File: proto-public/pbmesh/v1alpha1/computed_routes.proto
Comment (rboyer) [inline proto-public/pbmesh/v1alpha1/computed_routes.proto:133]: These are validated at admission time to force them to be `nil`. These are here so that during the sidecar controller resource assembly we can park data for the ultimate backend targets next to the rest of the data for them without introducing a second parallel map.

File: internal/mesh/internal/types/xroute.go
Comment (rboyer) [inline internal/mesh/internal/types/xroute.go:278]: Now it's a `*uint32` so it can't be negative.

File: internal/mesh/internal/types/tcp_route.go
Comment (rboyer) [inline internal/mesh/internal/types/tcp_route.go:53]: Since each rule has no "match criteria" only the first one ever does anything.

File: internal/mesh/internal/types/intermediate/types.go
Comment (rboyer) [inline internal/mesh/internal/types/intermediate/types.go:63]: These fields are related to the cluster generation, which for something with xRoutes may have more than one, hence needing to move this data down into the `ComputedPortRoutes.Targets` field.

File: internal/mesh/internal/mappers/sidecarproxymapper/service_endpoints_mapper.go
Comment (rboyer) [inline internal/mesh/internal/mappers/sidecarproxymapper/service_endpoints_mapper.go:21]: I think I still need to do this.

File: internal/mesh/internal/mappers/sidecarproxymapper/computed_routes_mapper.go
Comment (rboyer) [inline internal/mesh/internal/mappers/sidecarproxymapper/computed_routes_mapper.go:42]: I copied this comment from another place because the justification is still the same.

File: internal/mesh/internal/controllers/sidecarproxy/fetcher/data_fetcher.go
Comment (rboyer) [inline internal/mesh/internal/controllers/sidecarproxy/fetcher/data_fetcher.go:118]: This is the service with the same name as the one pointed to by the Upstreams item. It is the "logical" service you are trying to reach, **before** xRoute scrambling.

It no longer makes sense to examine `ServiceEndpoints` at  this layer, as that applies to the "real" services you reach on the other end of the xRoutes.

File: agent/xdsv2/listener_resources.go
Comment (rboyer) [inline agent/xdsv2/listener_resources.go:361]: Accounting for a TCPRoute split.

File: command/resource/apply/apply.go
Comment (rboyer) [inline command/resource/apply/apply.go:2]: `copywrite headers` fixup in passing

File: internal/mesh/internal/controllers/routes/generate.go
Comment (rboyer) [inline internal/mesh/internal/controllers/routes/generate.go:246]: GAMMA basically implies that "first xRoute written wins", so in that case whatever protocol the user has defined on the `Service` is not the _effective_ protocol in use when we render the xDS.

We ultimately need the rendered xDS to be self-consistent, so when we have an `HTTPRoute` with an underlying `Protocol` of `"tcp"` we upgrade it when we pass it off to the proxy state template layer.

File: internal/mesh/internal/controllers/routes/routestest/routestest.go
Comment (rboyer) [inline internal/mesh/internal/controllers/routes/routestest/routestest.go:3]: This package is meant to be helper code to facilitate turning a pile of resources into a `ComputedRoutes` resource without requiring the author of a test to manually populate the fields.

File: internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go
Comment (rboyer) [inline internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go:39]: I've flattened the implicit/explicit branches due them being so similar after my refactor it felt like a bad idea to duplicate so much code to maintain the split.

File: internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go
Comment (rboyer) [inline internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go:51]: Each key in the computed routes targets map will turn into a Cluster (or after failover works, an aggregate cluster).

File: internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go
Comment (rboyer) [inline internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go:284]: BUG: This needs to use the `meshPortName` of the target, not the upstream.

File: internal/mesh/internal/controllers/sidecarproxy/controller.go
Comment (rboyer) [inline internal/mesh/internal/controllers/sidecarproxy/controller.go:40]: TODO: need watch on `Service` now.

File: internal/mesh/internal/controllers/sidecarproxy/fetcher/data_fetcher.go
Comment (rboyer) [inline internal/mesh/internal/controllers/sidecarproxy/fetcher/data_fetcher.go:110]: likely need to `continue` if `d.Explicit` returns `nil` due to cache inconsistency.

File: internal/mesh/internal/types/intermediate/types.go
Comment (ishustava) [inline internal/mesh/internal/types/intermediate/types.go:61]: I think we can remove these as there are no other PRs stacked on top currently (or at least nothing big)

File: proto-public/pbmesh/v1alpha1/computed_routes.proto
Comment (ishustava) [inline proto-public/pbmesh/v1alpha1/computed_routes.proto:44]: wondering if we can just remove this field because we didn't ship it yet

**Review by ishustava — APPROVED** [review]
A few comments mostly about cleanup but nothing blocking so approving assuming these are addressed before merging.

File: internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go
Comment (jmurret) [reply internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go:70]: I wrote this and you addressed that you set this for future proofing.  If that's the case, do you want to add a comment.  I would have come along and deleted this.

**Review by jmurret — APPROVED** [review]
Nice work.  My only comments were related to code revisions for future tasks that can be done by any of us in the code.  🎉

File: internal/mesh/internal/cache/sidecarproxycache/computed_routes_cache.go
Comment (rboyer) [inline internal/mesh/internal/cache/sidecarproxycache/computed_routes_cache.go:1]: This is new from the initial review.

File: internal/mesh/internal/mappers/sidecarproxymapper/computed_routes_mapper.go
Comment (rboyer) [inline internal/mesh/internal/mappers/sidecarproxymapper/computed_routes_mapper.go:1]: All of the code in this package was changed/updated after the initial review.

File: internal/mesh/internal/controllers/routes/generate.go
Comment (rboyer) [inline internal/mesh/internal/controllers/routes/generate.go:295]: This now grabs the mesh port during ComputedRoutes construction so that the sidecar controller can skip resolving the Service resource just to get this data out again.

File: internal/mesh/internal/controllers/sidecarproxy/builder/naming.go
Comment (rboyer) [reply internal/mesh/internal/controllers/sidecarproxy/builder/naming.go]: no

File: internal/mesh/internal/controllers/sidecarproxy/builder/testdata/destination/l4-implicit-and-explicit-destinations-tproxy.golden
Comment (rboyer) [inline internal/mesh/internal/controllers/sidecarproxy/builder/testdata/destination/l4-implicit-and-explicit-destinations-tproxy.golden:119]: I guess this was a bug I fixed.

File: internal/mesh/internal/controllers/sidecarproxy/controller.go
Comment (rboyer) [reply internal/mesh/internal/controllers/sidecarproxy/controller.go:87]: I needed it for my changes unfortunately.

File: internal/mesh/internal/controllers/sidecarproxy/fetcher/data_fetcher.go
Comment (rboyer) [inline internal/mesh/internal/controllers/sidecarproxy/fetcher/data_fetcher.go:261]: I suspect this was a subtle bug before, having the explicit/implicit de-dupe map not include port.

File: internal/mesh/internal/controllers/sidecarproxy/fetcher/data_fetcher.go
Comment (rboyer) [reply internal/mesh/internal/controllers/sidecarproxy/fetcher/data_fetcher.go:494]: Yeah I missed this half-refactor. It's dead now.

File: internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go
Comment (rboyer) [reply internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go]: It's been refactored to avoid looking like it's doing nothing now.

**Pull request merged**
